sklearn/utils/validation.py

```diff
@@ -982,3 +982,121 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
 
     if max_val is not None and x > max_val:
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
+
+
+def check_kernel_eigenvalues(lambdas, warn_on_zeros=False):
```

Comment:
`utils.validation.py` is used for input checking so I don't think this belongs here. It should probably be in `kernel_pca.py` instead.

Also it should be private: users shouhldn't be allowed to use it directly (nor should they need to).

Reply:
I see. Note that `check_kernel_eigenvalues` is something that could be used everywhere where a kernel spectrum is computed, whether using exact or approximation methods. Or more generally a semi positive definite matrix's spectrum. So I thought that I would put it somewhere more central so that other developers in sklearn get a chance to try to reuse it. I thought that `utils.validation.py` was a good place but apparently I was wrong, where would be a better place for common internal validation tools? Thanks !

Reply:
I don't mind this here so much...

Reply:
Ok, then we might want to update the current module's docstring which says 
`"""Utilities for input validation"""`.

@smarie, since the function is relevant for any semi positive definite matrix, maybe change the name to e.g. (sorry, long name) `check_semi_positive_definite_eigenvalues()`

Also, would it be relevant to use it in the `PCA` code as well?

Reply:
If size matters, `check_spd_eigenvalues()` is almost as understandable.
If we go in that direction, I'll have to rename `KernelWarning` into `SpdSpectrumWarning`.

It does indeed apply to `PCA` and in general to all places where we use `eig` (non-symmetric matrix) or `eigh` (symmetric matrix) to compute the spectrum of a SPD matrix. I would be surprised that we use `eig` anywhere, but just in case...

Reply:
note: `SPDSpectrumWarning` would look better than `SpdSpectrumWarning`